distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Base_Generator.enso

```diff
@@ -45,6 +47,241 @@ type Dialect_Operations
     is_supported self operation =
         self.operations_dict.contains_key operation
 
+## Private
+
+   Generates SQL from a `Query`.
+
+   SQL generation requires some state threaded through the generation process,
+   and that state is stored in the fields of this type. This type only contains
+   the mutually recursive generators for the SQL IR that require this state.
+type SQL_Generator
+    ## We can use a Hashset for the let-binding names, since they are guaranteed
+       to be distinct and not result in any shadowing.
+    Value (let_bindings_ref : Ref Hashset)
+
+    new -> SQL_Generator = SQL_Generator.Value (Ref.new Hashset.empty)
+
+    ## PRIVATE
+
+       Builds code for an expression.
+
+       Arguments:
+       - dialect: The SQL dialect in which the expression is being generated.
+       - expr: The expression to generate SQL code for.
+    generate_expression : Dialect -> SQL_Expression | Order_Descriptor | Query -> SQL_Builder
+    generate_expression self dialect expr = case expr of
+        SQL_Expression.Column origin name ->
+            dialect.wrap_identifier origin ++ '.' ++ dialect.wrap_identifier name
+        SQL_Expression.Constant value -> SQL_Builder.interpolation value
+        SQL_Expression.Literal value -> SQL_Builder.code value
+        SQL_Expression.Text_Literal value ->
+            escaped = value.replace "'" "''"
+            SQL_Builder.code ("'" + escaped + "'")
+        SQL_Expression.Operation kind arguments metadata ->
+            op = dialect.dialect_operations.operations_dict.get kind (Error.throw <| Unsupported_Database_Operation.Error kind)
+            parsed_args = arguments.map (self.generate_expression dialect)
+            result = op parsed_args
+            # If the function expects more arguments, we pass the metadata as the last argument.
+            case result of
+                _ : Function -> result metadata
+                _ -> result
+        SQL_Expression.Let _ binder bindee body ->
+            wrapped_binder = dialect.wrap_identifier binder
+            bindee_exp = SQL_Builder.code "SELECT " ++ (self.generate_expression dialect bindee).paren ++ " AS " ++ "x"
+
+            body_sql_builder = self.let_bindings_ref.with_modification (h-> h.insert binder) <|
+                self.generate_expression dialect body
+
+            body_exp = SQL_Builder.code "SELECT " ++ body_sql_builder ++ " FROM " ++ wrapped_binder
+            (SQL_Builder.code "WITH " ++ wrapped_binder  ++ " AS " ++ bindee_exp.paren ++ " " ++ body_exp).paren
+        SQL_Expression.Let_Ref _ binder standalone_expression ->
+            binding_state = self.let_bindings_ref.get
+            case binding_state.contains binder of
+                True ->
+                    wrapped_binder = dialect.wrap_identifier binder
+                    wrapped_binder ++ "." ++ "x"
+                False ->
+                    self.generate_expression dialect standalone_expression
```

Comment:
Are there any other significant modifications to `Base_Generator` apart from this one + adding the `Ref`-based state?

The diff is huge 😅

Reply:
Found `preprocess_query` and siblings. Did I miss anything else that could be important?

Reply:
I think that nothing else changed other than handling the two new `SQL_Expression` constructors.

Reply:
Yes, confirmed, no other changes.